Sources/HTTPParser/http_parser.swift

```diff
@@ -61,20 +173,44 @@ private let HTTP_MAX_HEADER_SIZE = 80*1024
  * many times for each string. E.G. you might get 10 callbacks for "on_url"
  * each providing just a few characters more data.
  */
-public protocol http_parser_delegate {
-  func on_message_begin() -> Int
-  func on_url(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_status(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_header_field(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_header_value(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_headers_complete() -> Int
-  func on_body(at: UnsafePointer<UInt8>, length: Int) -> Int
-  func on_message_complete() -> Int
+public struct http_parser_delegate {
+  let on_message_begin: () -> Int
+  let on_url: ( UnsafePointer<UInt8>, Int) -> Int
+  let on_status: ( UnsafePointer<UInt8>, Int) -> Int
+  let on_header_field: ( UnsafePointer<UInt8>, Int) -> Int
+  let on_header_value: ( UnsafePointer<UInt8>, Int) -> Int
+  let on_headers_complete: () -> Int
+  let on_body: ( UnsafePointer<UInt8>, Int) -> Int
+  let on_message_complete: () -> Int
   /* When on_chunk_header is called, the current chunk length is stored
    * in parser.content_length
    */
-  func on_chunk_header() -> Int
-  func on_chunk_complete() -> Int
+  let on_chunk_header: () -> Int
+  let on_chunk_complete: () -> Int
+
+  public init(on_message_begin: @escaping () -> Int,
```

Comment:
In the original the parser itself doesn't maintain the callbacks, it is kept in a separate struct which is passed to execute(). This makes sense because the parser is reentrant/stateless and the mapping is usually static (only one copy needed). Remember that the idea is to keep the per-connection overhead as small as possible (in the original only 40 bytes per connection!).